poseidon2-circuit-air/src/air.rs

```diff
@@ -0,0 +1,622 @@
+
+        let mut state = [F::ZERO; WIDTH];
+        let mut inputs = Vec::with_capacity(n);
+        for (i, op) in sponge_ops.iter().enumerate() {
+            let Poseidon2CircuitRow {
+                is_sponge,
+                reset,
+                absorb_flags,
+                input_values,
+                input_indices,
+                output_indices,
+            } = op;
+
+            let row = circuit_trace.row_mut(i);
+
+            row[0] = F::from_bool(*is_sponge);
+            row[1] = F::from_bool(*reset);
+            row[2] = F::from_bool(*is_sponge && *reset);
+            for j in 0..RATE_EXT {
+                row[3 + j] = F::from_bool(absorb_flags[j]);
+            }
+            for j in 0..RATE_EXT {
+                row[3 + RATE_EXT + j] = F::from_u32(input_indices[j]);
+            }
+            for j in 0..RATE_EXT {
+                row[3 + RATE_EXT + WIDTH_EXT + j] = F::from_u32(output_indices[j]);
+            }
+
+            let mut index_absorb = [false; RATE_EXT];
+            for (j, flag) in absorb_flags.iter().enumerate() {
+                if *flag {
+                    for absorb in index_absorb.iter_mut().take(j + 1) {
+                        *absorb = true;
+                    }
+                }
+            }
+
+            for (j, absorb) in index_absorb.iter_mut().enumerate() {
+                if *absorb {
+                    for d in 0..D {
+                        let idx = j * D + d;
+                        state[idx] = input_values[idx];
+                    }
+                } else if *reset {
+                    // During a reset, non-absorbed rate elements are zeroed.
+                    for d in 0..D {
+                        let idx = j * D + d;
+                        state[idx] = F::ZERO;
+                    }
+                }
+            }
+
+            if *reset || !*is_sponge {
+                // Compression or reset: reset capacity
+                for j in 0..(CAPACITY_EXT * D) {
+                    state[RATE_EXT * D + j] = F::ZERO;
+                }
+            }
+
+            inputs.push(state);
+            state = perm.permute(state);
+        }
+
+        let p2_trace = generate_trace_rows::<
+            F,
+            LinearLayers,
+            WIDTH,
+            SBOX_DEGREE,
+            SBOX_REGISTERS,
+            HALF_FULL_ROUNDS,
+            PARTIAL_ROUNDS,
+        >(inputs, constants, extra_capacity_bits);
+
+        let ncols = self.width();
+
+        let p2_ncols = p3_poseidon2_air::num_cols::<
+            WIDTH,
+            SBOX_DEGREE,
+            SBOX_REGISTERS,
+            HALF_FULL_ROUNDS,
+            PARTIAL_ROUNDS,
+        >();
+
+        debug_assert_eq!(ncols, num_circuit_cols + p2_ncols);
+
+        let mut vec = vec![F::ZERO; n * ncols];
+
+        let circuit_trace_view = circuit_trace.as_view();
+
+        for ((row, left_part), right_part) in vec
```

Comment:
Here we need to add TODO for removing copying the already computed `p2_trace`

Reply:
Actually, this would be easy if [`p3_poseidon2_air::generate_trace_rows_for_perm`](https://github.com/Plonky3/Plonky3/blob/7656e166e044cb7955b783ce4108e87e1cecf47e/poseidon2-air/src/generation.rs#L134) becomes public. We would just need to declare the trace matrix as
```rust
let num_permutation_cols = p3_poseidon2_air::num_cols::<
            WIDTH,
            SBOX_DEGREE,
            SBOX_REGISTERS,
            HALF_FULL_ROUNDS,
            PARTIAL_ROUNDS,
        >();
let mut circuit_trace = vec![F::ZERO; n * (num_circuit_cols + num_permutation_cols)];
```
and write each poseidon2 permutation row, using `p3_poseidon2_air::generate_trace_rows_for_perm`, at the end of each row.